### PR TITLE
Prefix Composer Dependencies

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

Prefixes dependencies to avoid conflicts with the AMP plugin and other plugins using these dependencies.

## Relevant Technical Choices

* Uses [PHP-Scoper](https://github.com/humbug/php-scoper) to prefix dependencies  
  * Installs it via `civicrm/composer-downloads-plugin` to work around PHP version mismatch issues
  * Updates CI config accordingly
* Separating autoloaders in `third-party` and `includes` makes building the plugin ZIP file much faster
* Removes `composer validate` CI step as it doesn't work with `mcaskill/composer-exclude-files`

## To-do

* [x] Investigate potential `ClassLoader` compatibility issues  
  See https://github.com/humbug/php-scoper/issues/355, https://github.com/composer/composer/issues/3852, https://github.com/dangoodman/composer-for-wordpress, https://gist.github.com/polevaultweb/de44d747b57191a851947d5f5175e922

## User-facing changes

N/A

## Testing Instructions

1. Check out this branch and run `composer install` _or_ check out this branch on the QA site
1. Activate AMP plugin
1. Visit a single story, edit a story, etc.
1. Verify there are no PHP errors or notices

---

Addresses one aspect of #4513

### CODE REVIEW

File: bin/utils/createBuild.js [hunk 1]
```diff
@@ -14,20 +14,12 @@
...
- */
-import { execSync } from 'child_process';
-
 /**
  * Internal dependencies
  */
 import copyFiles from './copyFiles.js';
 import getIgnoredFiles from './getIgnoredFiles.js';
 
-const composerArgs =
```
Comment (spacedmonkey) [inline bin/utils/createBuild.js:28]: Why was this removed?

File: composer.json [hunk 2]
```diff
@@ -32,6 +34,15 @@
     },
     "sort-packages": true
   },
+  "extra": {
```
Comment (spacedmonkey) [inline composer.json]: This should be installed via packagist. https://packagist.org/packages/humbug/php-scoper

See the comment. https://github.com/google/web-stories-wp/pull/4540/files#r491043156

Either use `--ignore-platform-reqs` flag or change to the platform to 7.2. See [example](https://github.com/xwp/unsplash-wp/blob/develop/composer.json#L9-L13) from unsplash.

File: bin/utils/createBuild.js (hunk 1 above)
Comment (swissspidy) [reply bin/utils/createBuild.js:28]: Because it's no longer needed

File: composer.json (hunk 2 above)
Comment (swissspidy) [reply composer.json]: We cannot change the platform to 7.2 as the plugin supports WordPress 5.6.

As I mentioned in the other thread, `--ignore-platform-reqs` only has downsides here.

PHP Scoper itself **recommends** using the PHAR installation method, see https://github.com/humbug/php-scoper#phar. Their example uses bamarni/composer-bin-plugin, which is similar to the one I use here.

Comment (spacedmonkey) [reply composer.json]: But if I install this locally and want to contribute and I am running say 7.1, then this will not work.

Comment (swissspidy) [reply composer.json]: For contributors we recommend a minimum required PHP version, just like we recommend the Node version.

It‘s reasonable to expect contributors to not run an outdated version of PHP IMO

Comment (spacedmonkey) [reply composer.json]: So what is the issue of including this as a composer dev dependency?

Comment (swissspidy) [reply composer.json]: Exactly the same reason as #4540

**Comment by swissspidy** [comment]
@westonruter @schlessera @spacedmonkey  I feel like this is now ready for initial review, so I'd love to get your feedback. Please note that this PR is based on #4513 and solves the conflict resolving part of that PR.

@felixarntz FYI I added you as a reviewer in case you have some feedback for us on this based on prior art in Site Kit. The Site Kit implementation was certainly a good inspiration, although I went for an easier route heree.

File: composer.json [hunk 3]
```diff
@@ -20,6 +20,7 @@
   },
   "require-dev": {
     "automattic/vipwpcs": "^2.1",
+    "civicrm/composer-downloads-plugin": "^2.1",
```
Comment (felixarntz) [inline composer.json:24]: Nice! I assume this is to install this regardless of PHP version set in `composer.json`? Maybe worth for us to change that in Site Kit too.

File: composer.json [hunk 4]
```diff
@@ -64,5 +72,20 @@
...
+  "prefer-stable": true,
+  "scripts": {
+    "post-install-cmd": [
+      "@prefix-dependencies"
+    ],
+    "post-update-cmd": [
+      "@prefix-dependencies"
+    ],
+    "prefix-dependencies": [
+      "php-scoper add-prefix --output-dir=./third-party --force --quiet",
```
Comment (felixarntz) [inline composer.json:94]: We're using `php-scoper add ...` - is that an alias or maybe something changed in a newer version?

File: composer.json [hunk 5]
```diff
@@ -64,5 +72,20 @@
...
+      "echo '{ \"autoload\": { \"classmap\": [\"\"] } }' > ./third-party/composer.json",
+      "@composer dump-autoload --working-dir ./third-party --no-dev --classmap-authoritative",
+      "echo '{ \"autoload\": { \"classmap\": [\"\"] } }' > ./includes/composer.json",
+      "@composer dump-autoload --working-dir ./includes --no-dev --classmap-authoritative"
```
Comment (felixarntz) [inline composer.json]: In Site Kit we only rely on the actual classmap, not the full Composer autoloader. Which is certainly less elegant and a more manual approach, but there's still some chance for conflicts when relying on the Composer autoloader, e.g. due to its `Composer\Autoload\ClassLoader` class which could exist in multiple projects in different versions.

File: scoper.inc.php [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+	// Paths are relative to the configuration file unless if they are already absolute.
+	'files-whitelist'            => [],
+
+	// When scoping PHP files, there will be scenarios where some of the code being scoped indirectly references the
+	// original namespace. These will include, for example, strings or string manipulations. PHP-Scoper has limited
+	// support for prefixing such strings. To circumvent that, you can define patchers to manipulate the file to your
+	// heart contents.
+	//
+	// For more see: https://github.com/humbug/php-scoper#patchers.
+	'patchers'                   => [
```
Comment (felixarntz) [inline scoper.inc.php:151]: OMG. Now I feel lucky Site Kit doesn't use WordPress plugins as dependency - class-based libraries work way better with this than global WordPress function mess 😆

File: composer.json (hunk 3 above)
Comment (swissspidy) [reply composer.json:24]: Yep! :-)

File: composer.json (hunk 4 above)
Comment (swissspidy) [reply composer.json:94]: `add` is just a shorthand alias for `add-prefix` 👍

File: scoper.inc.php (hunk 6 above)
Comment (swissspidy) [reply scoper.inc.php:151]: 😅 yes you're lucky!

I was hoping the [whitelist options](https://github.com/humbug/php-scoper#whitelist) were a bit better. Maybe I need to give them another try though.

File: composer.json (hunk 5 above)
Comment (swissspidy) [reply composer.json]: Interesting, just stumbled upon https://github.com/composer/composer/issues/3852 that discusses this.

https://github.com/dangoodman/composer-for-wordpress is mentioned there as an alternative to rolling one's own autoloader. I'm gonna give it a try.

Comment (swissspidy) [reply composer.json]: @schlessera Also curious to hear your thoughts/ideas on this

Comment (swissspidy) [reply composer.json]: Went with a simple solution to just prefix the `ClassLoader` namespace. Way easier than all the other methods.

**Review by westonruter — APPROVED** [review]
I've never used php-scoper before, but LGTM.
